File: backend/app/esquemas/reportes.py
import uuid
from datetime import datetime
from typing import Generic, TypeVar

from pydantic import BaseModel

from app.esquemas.catalogo import DineroSalida
from app.esquemas.compras import ProductoBreve, ProveedorBreve

T = TypeVar("T")


class Lista(BaseModel, Generic[T]):  # noqa: UP046  (Pydantic y PEP 695)
    datos: list[T]
    cursor_siguiente: str | None
    tiene_mas: bool


class CategoriaBreve(BaseModel):
    id: uuid.UUID
    nombre: str


class FilaBajoMinimo(BaseModel):
    producto: ProductoBreve
    stock_actual: str
    stock_minimo: str
    deficit: str
    deficit_relativo: str


class FilaAgotado(BaseModel):
    producto: ProductoBreve
    stock_actual: str
    stock_minimo: str | None


class FilaSinMovimiento(BaseModel):
    producto: ProductoBreve
    stock_actual: str
    valor_a_costo: DineroSalida | None
    ultimo_movimiento_en: datetime | None
    creado_en: datetime


class ValorCategoria(BaseModel):
    categoria: CategoriaBreve | None
    productos: int
    valor: DineroSalida


class FilaNoValorizable(BaseModel):
    producto: ProductoBreve
    stock_actual: str


class Valorizacion(BaseModel):
    """RF-REP-003: stock × costo actual en moneda base (RN-09). Sin parámetro de fecha."""

    total: DineroSalida
    productos_valorizados: int
    por_categoria: list[ValorCategoria]
    no_valorizables: Lista[FilaNoValorizable]


class ComprasProveedor(BaseModel):
    proveedor: ProveedorBreve
    total_recibido: DineroSalida
    total_facturado: DineroSalida


class ComprasCategoria(BaseModel):
    categoria: CategoriaBreve | None
    total_recibido: DineroSalida


class ResumenCompras(BaseModel):
    """RF-REP-005: en moneda base con las tasas congeladas en cada documento."""

    desde: str
    hasta: str
    total_recibido: DineroSalida
    total_facturado: DineroSalida
    recepciones: int
    facturas: int
    por_proveedor: list[ComprasProveedor]
    por_categoria: list[ComprasCategoria]


class MermaMotivo(BaseModel):
    motivo: str
    etiqueta: str
    cantidad: str
    valor: DineroSalida


class MermaProducto(BaseModel):
    producto: ProductoBreve
    cantidad: str
    valor: DineroSalida


class ResumenMermas(BaseModel):
    """RF-REP-006 / RN-16: solo tipo merma, sin anuladas, valorizadas al costo actual."""

    desde: str
    hasta: str
    total_cantidad: str
    total_valor: DineroSalida
    por_motivo: list[MermaMotivo]
    por_producto: Lista[MermaProducto]


class FilaDiscrepancia(BaseModel):
    movimiento_id: uuid.UUID
    producto: ProductoBreve
    tipo: str
    cantidad: str
    stock_resultante: str
    motivo: str
    nota: str | None
    ocurrido_en: datetime
    autor_tipo: str

File: android/core/domain/src/main/kotlin/co/inventario/domain/modelo/Reportes.kt
package co.inventario.domain.modelo

import java.time.Instant

/** RF-REP-001: ordenado por criticidad (déficit relativo al mínimo). */
data class FilaBajoMinimo(val producto: ProductoBreve, val stockActual: Cantidad, val stockMinimo: Cantidad, val deficit: Cantidad, val deficitRelativo: String)

/** RF-REP-007. */
data class FilaAgotado(val producto: ProductoBreve, val stockActual: Cantidad, val stockMinimo: Cantidad?)

/** RF-REP-002. */
data class FilaSinMovimiento(val producto: ProductoBreve, val stockActual: Cantidad, val valorACosto: Dinero?, val ultimoMovimientoEn: Instant?, val creadoEn: Instant)

data class ValorCategoria(val categoria: Categoria?, val productos: Int, val valor: Dinero)

data class FilaNoValorizable(val producto: ProductoBreve, val stockActual: Cantidad)

/** RF-REP-003: los no valorizables van aparte, nunca como cero. */
data class Valorizacion(val total: Dinero, val productosValorizados: Int, val porCategoria: List<ValorCategoria>, val noValorizables: Pagina<FilaNoValorizable>)

data class ComprasProveedor(val proveedor: ProveedorBreve, val totalRecibido: Dinero, val totalFacturado: Dinero)

data class ComprasCategoria(val categoria: Categoria?, val totalRecibido: Dinero)

/** RF-REP-005. */
data class ResumenCompras(
    val desde: String,
    val hasta: String,
    val totalRecibido: Dinero,
    val totalFacturado: Dinero,
    val recepciones: Int,
    val facturas: Int,
    val porProveedor: List<ComprasProveedor>,
    val porCategoria: List<ComprasCategoria>,
)

data class MermaMotivo(val motivo: String, val etiqueta: String, val cantidad: Cantidad, val valor: Dinero)

data class MermaProducto(val producto: ProductoBreve, val cantidad: Cantidad, val valor: Dinero)

/** RF-REP-006 / RN-16. */
data class ResumenMermas(
    val desde: String,
    val hasta: String,
    val totalCantidad: Cantidad,
    val totalValor: Dinero,
    val porMotivo: List<MermaMotivo>,
    val porProducto: Pagina<MermaProducto>,
)

/** RF-INV-006: movimientos forzados, consultables como discrepancias. */
data class FilaDiscrepancia(
    val movimientoId: String,
    val producto: ProductoBreve,
    val tipo: TipoMovimiento,
    val cantidad: Cantidad,
    val stockResultante: Cantidad,
    val motivo: String,
    val nota: String?,
    val ocurridoEn: Instant,
    val autorTipo: String,
)
